synapse/handlers/oidc_handler.py

```diff
@@ -0,0 +1,452 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 Quentin Gliech
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+from typing import Dict, List, Optional
+from urllib.parse import parse_qsl
+
+import pymacaroons
+from authlib.common.security import generate_token
+from authlib.jose import JsonWebToken
+from authlib.oauth2.auth import ClientAuth
+from authlib.oauth2.rfc6749.parameters import prepare_grant_uri
+from authlib.oidc.core import CodeIDToken, ImplicitIDToken, UserInfo
+from authlib.oidc.discovery import OpenIDProviderMetadata, get_well_known_url
+from jinja2 import Template
+from pymacaroons.exceptions import MacaroonDeserializationException
+
+from synapse.api.errors import HttpResponseException
+from synapse.http.server import finish_request
+from synapse.push.mailer import load_jinja2_templates
+from synapse.types import UserID, map_username_to_mxid_localpart
+
+logger = logging.getLogger(__name__)
+
+SESSION_COOKIE_NAME = b"oidc_session"
+
+
+class OidcError(Exception):
```

Comment:
Should this inherit from `SynapseError`?

Reply:
I don't think so: it is only raised in `_exchange_code` to then be caught in `handle_oidc_callback` so we can display the error nicely. Not sure if it's the right approach though

Reply:
Ah, I think I missed where this is a caught, I'll need to look again! Thanks!